store/tikv/txn_committer.go

```diff
@@ -84,51 +84,70 @@ func (c *txnCommitter) primary() []byte {
 	return c.keys[0]
 }
 
-func (c *txnCommitter) iterKeysByRegion(keys [][]byte, f func(RegionVerID, [][]byte) error) error {
-	groups := make(map[RegionVerID][][]byte)
-	var primaryRegionID RegionVerID
-	var lastRegion *Region
-	for _, k := range keys {
-		var region *Region
-		if lastRegion != nil && lastRegion.Contains(k) {
-			region = lastRegion
-		} else {
-			var err error
-			region, err = c.store.regionCache.GetRegion(k)
-			if err != nil {
-				return errors.Trace(err)
-			}
-			lastRegion = region
-		}
-		id := region.VerID()
-		if bytes.Compare(k, c.primary()) == 0 {
-			primaryRegionID = id
-		}
-		groups[id] = append(groups[id], k)
+// iterKeys groups keys into batches, then applies `f` to the them. If the flag
+// asyncNonPrimary is set, it will return as soon as the primary batch is
+// processed.
+func (c *txnCommitter) iterKeys(keys [][]byte, f func(batchKeys) error, sizeFn func([]byte) int, asyncNonPrimary bool) error {
+	if len(keys) == 0 {
+		return nil
 	}
+	groups, firstRegion, err := c.store.regionCache.GroupKeysByRegion(keys)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	firstIsPrimary := bytes.Equal(keys[0], c.primary())
 
+	var batches []batchKeys
 	// Make sure the group that contains primary key goes first.
-	if primaryRegionID.id != 0 {
-		if err := f(primaryRegionID, groups[primaryRegionID]); err != nil {
+	if firstIsPrimary {
+		batches = appendBatchBySize(batches, firstRegion, groups[firstRegion], sizeFn)
+		delete(groups, firstRegion)
+	}
+	for id, g := range groups {
+		batches = appendBatchBySize(batches, id, g, sizeFn)
+	}
+
+	if firstIsPrimary {
+		err = c.doBatches(batches[:1], f)
+		if err != nil {
 			return errors.Trace(err)
 		}
-		delete(groups, primaryRegionID)
+		batches = batches[1:]
 	}
-	// Parallel all rest Regions.
-	// TODO: Stop sending other requests after receiving first error.
-	ch := make(chan error)
-	for id, g := range groups {
-		go func(id RegionVerID, g [][]byte) {
-			ch <- f(id, g)
-		}(id, g)
+	if asyncNonPrimary {
+		go c.doBatches(batches, f)
+		return nil
 	}
+	err = c.doBatches(batches, f)
+	return errors.Trace(err)
+}
+
+// doBatches applies f to batches parallelly.
+func (c *txnCommitter) doBatches(batches []batchKeys, f func(batchKeys) error) error {
 	var err error
-	for i := 0; i < len(groups); i++ {
-		if e := <-ch; e != nil {
-			err = errors.Trace(e)
+	if len(batches) == 0 {
+		return nil
+	} else if len(batches) == 1 {
+		err = errors.Trace(f(batches[0]))
+		if err != nil {
+			log.Warnf("txnCommitter doBatches failed: %v, tid: %d", err, c.startTS)
+		}
+	} else {
+		// TODO: Stop sending other requests after receiving first error.
```

Comment:
@coocood We may do it for `prewrite`. In prewrite phase, conflict in any single region will abort the whole txn, not matter it contains primary key or not.